### PR TITLE
Support nested types in `lists::contains`

### PR DESCRIPTION
**Pull request by ttnghia**
This extends the `lists::contains` API to support nested types (lists + structs) with arbitrarily nested levels. As such, `lists::contains` will work with literally any type of input data.

In addition, the related implementation has been significantly refactored to facilitate adding new implementation.

Closes https://github.com/rapidsai/cudf/issues/8958.
Depends on:
 * https://github.com/rapidsai/cudf/pull/10730
 * https://github.com/rapidsai/cudf/pull/10883
 * https://github.com/rapidsai/cudf/pull/10999
 * https://github.com/rapidsai/cudf/pull/11019
 * https://github.com/rapidsai/cudf/issues/11037

### CODE REVIEW

File: cpp/include/cudf/lists/lists_column_device_view.cuh
Comment (jrhemstad) [inline cpp/include/cudf/lists/lists_column_device_view.cuh:25]: Why was this removed? Is this now intended to be a publicly consumable type?

Comment (ttnghia) [reply cpp/include/cudf/lists/lists_column_device_view.cuh:25]: I observed that most other `*_device_view` classes are not in the `detail::` namespace thus want to make things more consistent.

Comment (jrhemstad) [reply cpp/include/cudf/lists/lists_column_device_view.cuh:25]: There are many things that are not in a detail namespace that should be. This is moving in the wrong direction.

Comment (ttnghia) [reply cpp/include/cudf/lists/lists_column_device_view.cuh:25]: Okay then I'll reverse it. Next time will move things into `detail::` instead.

File: cpp/src/lists/contains.cu
Comment (jrhemstad) [inline cpp/src/lists/contains.cu:357]: This flattening logic should not be necessary anymore with the new row_comparators.

Comment (ttnghia) [reply cpp/src/lists/contains.cu:357]: But it is still in `experimental::` namespace, isn't it?

Comment (ttnghia) [reply cpp/src/lists/contains.cu:357]: Struct flattening has already been used in many places (sorting/min/max/contains etc...). I planned to use this flattening old way for all the ongoing FEA (`lists::contains` and `cudf::contains`), then make a separate, big PR to convert all these to use the new experimental row comparator.

Comment (devavret) [reply cpp/src/lists/contains.cu:357]: The namespace is used for the purpose of coexisting the old comparators and the new comparators to facilitate a gradual transition. 

If the new row comparators will make this PR simpler while adding functionality, you can just use it without worrying much about stability.

Comment (ttnghia) [reply cpp/src/lists/contains.cu:357]: Okay got it. Thanks.

Comment (devavret) [reply cpp/src/lists/contains.cu:357]: Also, no need to make a large big bang PR that changes over all contains logic to use the new comparators. You can take it one step at a time. 

Since the original issue #8958 does request deeply nested structs, you would get that functionality for "free".

Comment (ttnghia) [reply cpp/src/lists/contains.cu:357]: So, based on our discussions, we still don't have eq comparator for 2 structs columns yet. I'll just use the old flattening way for this PR. A complete conversion to use the new comparator will be done later.

Comment (jrhemstad) [reply cpp/src/lists/contains.cu:357]: > I'll just use the old flattening way for this PR. A complete conversion to use the new comparator will be done later.

Let's not add throwaway work of building new functionality on top of something we know we want to get rid of.

Comment (ttnghia) [reply cpp/src/lists/contains.cu:357]: Okay. So @bdice I'm relying on you :smirk:

**Comment by ttnghia** [comment]
Currently hold off due to dependency.

**Comment by ttnghia** [comment]
We decided to retarget this into 22.08 to prioritize other more important FEA.

File: cpp/src/lists/contains.cu
Comment (ttnghia) [reply cpp/src/lists/contains.cu:188]: `idx` is the index of elements within a list. `list.row_index()` is the index of the list in the lists column.

Comment (ttnghia) [reply cpp/src/lists/contains.cu:188]: `lhs` is the child of the lists column (`n` lists), so we need to get the child row through `element_offset(idx)`. On the other hand, `rhs` is just the keys column having size `n` so we get the row through `list.row_index()`.

File: cpp/src/lists/contains.cu
Comment (mythrocks) [inline cpp/src/lists/contains.cu:292]: How come we're using return-type based SFINAE, instead of template parameter based?
```c++
template <typename Element, typename SearchKeyType, CUDF_ENABLE_IF(is_supported_type<Element>())>
std::unique_ptr<column> operator() (...) const {}
```
Have I asked this before?

Comment (ttnghia) [reply cpp/src/lists/contains.cu:292]: I added the comment lines just right above this.

File: cpp/src/lists/contains.cu
Comment (mythrocks) [inline cpp/src/lists/contains.cu:253]: Sorry, this part of it looks expensive. Could we not attempt this without having to blow up the scalar into a vector? The non-nested code path does not require this blow-up.

I would have thought this doable with an appropriate iterator.

File: cpp/src/lists/contains.cu
Comment (mythrocks) [reply cpp/src/lists/contains.cu:292]: Ah, right. I couldn't remember what the condition was that forces us to use return-type-based SFINAE.

File: cpp/src/lists/contains.cu
Comment (mythrocks) [reply cpp/src/lists/contains.cu:253]: > I would have thought this doable with an appropriate iterator.

Hmm... The row-comparator uses `table_view`, so this can't be done naively. We would need a table/column_view that mimics what the `optional_comparator` does for `scalar` values. It would be non-trivial.

I'm fine with this being merged, in the name of expediency. But I'd like for us to revisit this issue, to avoid having to blow up a scalar into a vector, simply because of an unfortunate interface.

Comment (ttnghia) [reply cpp/src/lists/contains.cu:253]: @GregoryKimball we talked about this before, when we want to compare a column (table) with a scalar but the existing comparator can only compare tables. Now we have more voice on this, so I think we should prioritize this somehow.

File: cpp/src/lists/contains.cu
Comment (bdice) [inline cpp/src/lists/contains.cu:185]: I'm not sure if we want to capture everything by reference here. The `list_device_view` should probably be passed by value unless you know of a reason not to do so.
```suggestion
      thrust::find_if(thrust::seq, begin, end, [list, &d_comp, &search_key_is_scalar] __device__(auto const idx) {
```

File: cpp/src/lists/contains.cu
Comment (bdice) [inline cpp/src/lists/contains.cu:205]: For symmetry (and generic iterators), I'd prefer to see a templated `OutputIterator` instead of `size_type*`.

File: cpp/src/lists/contains.cu
Comment (bdice) [inline cpp/src/lists/contains.cu:211]: Don't pass iterators by reference. This one's tricky, and is one of the reasons I prefer to always write my lambda captures out explicitly. 😉
```suggestion
  auto const do_search = [input_it, num_rows, output_it, find_option, stream](auto const keys_iter) {
```

Comment (bdice) [reply cpp/src/lists/contains.cu:211]: ... and similarly below, check your lambda captures and try to pass iterators and small fixed-width types (`int`, `bool`, enum classes) by value.

File: cpp/src/lists/contains.cu
Comment (bdice) [reply cpp/src/lists/contains.cu:205]: ...and similarly below in `index_of_nested_types`.

File: cpp/src/lists/contains.cu
Comment (bdice) [inline cpp/src/lists/contains.cu:237]: The naming here is asymmetric. Why `index_of_nested_types` and `index_of_non_nested`? Both (or neither) should use the suffix `_types`.

File: cpp/src/lists/contains.cu
Comment (bdice) [reply cpp/src/lists/contains.cu:253]: I think that comparators for columns/tables against scalars are a reasonable feature request. We can probably find a way to implement this efficiently with a wrapper of some kind around the experimental row operators. Can you file an issue, and list all the places where you know this would be useful?

Comment (bdice) [reply cpp/src/lists/contains.cu:253]: Upon further investigation: the current state is quite yucky, with lifetime management for one code path but not the other. We should be able to handle this a different way, by making a comparator that accepts a scalar, owns the temporary column, and accepts one index (for the column/table input). Centralizing lifetime management in the comparator has been part of our existing designs, so that's probably good to continue.

There aren't a lot of alternatives I can see without improving this at the comparator level, but it's pretty gross in the current state.

**Review by bdice — CHANGES_REQUESTED** [review]
Comments attached.

File: cpp/src/lists/contains.cu
Comment (bdice) [inline cpp/src/lists/contains.cu:435]: I'd like to see this split into two function calls on separate lines. It makes it less obvious how memory resources are used for intermediates when the functions are combined. Same below.

File: cpp/src/lists/contains.cu
Comment (bdice) [reply cpp/src/lists/contains.cu:258]: https://github.com/rapidsai/cudf/issues/10892#issuecomment-1217233559

@ttnghia If you can, please add any other cases to this issue where you know this problem is occurring. There are some alternatives I can imagine to the current behavior, but none of them are as good as solving the problem at the comparator level.

File: cpp/src/lists/contains.cu
Comment (bdice) [inline cpp/src/lists/contains.cu:102]:
```suggestion
  auto const begin = thrust::make_counting_iterator(0);
  auto const end   = thrust::make_counting_iterator(size);
  if constexpr (forward) {
    return thrust::pair(begin, end);
  } else {
    return thrust::pair(thrust::make_reverse_iterator(end), thrust::make_reverse_iterator(begin));
  }
```

File: cpp/src/lists/contains.cu
Comment (ttnghia) [reply cpp/src/lists/contains.cu:258]: > Can you file an issue, and list all the places where you know this would be useful?

Sorry I forgot to mention the already filed issue https://github.com/rapidsai/cudf/issues/10892. Thanks for bringing it up.

File: cpp/tests/lists/contains_tests.cpp
Comment (bdice) [inline cpp/tests/lists/contains_tests.cpp:104]: For the return values in these tests, we should check against values of `size_type` instead of hardcoding the present definition that `size_type = int32_t`.

**Review by bdice — CHANGES_REQUESTED** [review]
A few more comments -- I think I will approve when these are addressed. There's a lot going on in this PR and I've tried my best to make sense of it, but there's a lot of "patterning" with complex lambdas, boolean template dispatches, etc. I am struggling to find better recommendations but I don't think I can concretely identify what I would prefer in those cases, if anything, at this time.

**Review by mythrocks — APPROVED** [review]
LGTM!

File: cpp/tests/lists/contains_tests.cpp
Comment (bdice) [inline cpp/tests/lists/contains_tests.cpp:1105]: Are the trailing `//` necessary if clang-format is already turned off in this section? I'd remove the trailing `//`.

**Review by bdice — APPROVED** [review]
One minor comment, otherwise LGTM.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by ttnghia** [comment]
@gpucibot merge

**Pull request merged**
